lib/tokeniterators.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2021 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef tokeniterators_h__
+#define tokeniterators_h__
+
+
+#include "symboldatabase.h"
+#include "tokenlist.h"
+#include "tokenize.h"
+
+
+
+/**
+ * @brief  Syntax sugar to be used in c++11 range-based loops
+ */
+class IterateTokens {
+
+protected: // forbid to create iterator instances by user to avoid unexpected copy behavior

```

Comment:
Thats not a valid iterator, c++ iterators are required to be copy_constructible, copy_assignable, and swappable.